### PR TITLE
Tim/inbox

### PR DESCRIPTION
**Pull request by tgoodwin**

### CODE REVIEW

File: app/components/App.jsx [hunk 1]
```diff
@@ -15,7 +16,12 @@ class App extends React.Component {
 		// out normally
 		const unsubscribe = getAuth().onAuthStateChanged((user) => {
 		  if (user) {
-				this.props.fetchUser(user.email, user.uid).then(() => {
+		  		let tasks = [
+		  			this.props.fetchUser(user.email, user.uid),
+		  			this.props.getUserNotifications(user.uid)
+	  			];
+				Promise.all(tasks).then(() => {
```
Comment (tgoodwin) [inline app/components/App.jsx]: is there a better way to do this? something async and redux-y or something?

File: app/actions/inbox.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	notifications: data.entries,
+	nextPage: data.nextPage
+});
+
+export const getNotificationEventData = (notifications) => dispatch => {
+	dispatch(beginInboxFetch());
+	// can have multiple notifications on a given event.
+	// only fetch the bare minimum amt of data we need.
+	// also, only some notifications have postIds.
+	const uniquePostIds = _.uniq(_.pluck(notifications, 'postId')).filter(id => !!id);
```
Comment (tgoodwin) [inline app/actions/inbox.js]: just trying to get unique array of objects by a key here.. better way to do this with vanilla js?

File: app/actions/inbox.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	const queries = uniquePostIds.map(postId => {
+		return getPostData(postId).then(data => {
+			return {
+				...data.val(), postId
+			}
+		});
+	});
+	return Promise.all(queries).then(results => {
```
Comment (tgoodwin) [inline app/actions/inbox.js]: also I keep using `Promise.all(queries)` pattern but wondering if there is a better more Redux-y approach to firing off multiple async fetches

File: app/components/header/Header.jsx [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import { NavLink, Link } from 'react-router-dom';
+import InboxHeader from './InboxHeader';
+import SearchContainer from '../search/SearchContainer';
+import SearchToggle from '../search/SearchToggle';
+
+const Header = ({currentUser, logOut, loggedIn, unreadNotificationCount}) => {
+	const logOutBtn = (<button onClick={logOut}>Sign out</button>);
+	const logInBtn = (<NavLink to='/login'>Log in</NavLink>);
+	const inboxHeader = (
+		<InboxHeader inviteCount={3} />
```
Comment (tgoodwin) [inline app/components/header/Header.jsx]: wow oops

File: app/components/header/Header.jsx [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+						<SearchToggle />
+					</ul>
+					<SearchContainer />
+				</nav>
+			}
+		</div>
+	)
+	const isInbox = window.location.pathname === '/inbox';
+	return ( isInbox ? inboxHeader : defaultHeader)
+}
```
Comment (tgoodwin) [inline app/components/header/Header.jsx]: also this part uhhhh

File: app/components/inbox/Inbox.jsx [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+					key={notification.key}
+					id={notification.key}
+					postId={notification.postId}
+				/>
+			)
+		});
+	}
+	render() {
+		let notification = this.props.unreadNotificationCount == 1 ? 'notification' : 'notifications';
+		let count = this.props.unreadNotificationCount > 9 ? '9+' : this.props.unreadNotificationCount;
```
Comment (tgoodwin) [inline app/components/inbox/Inbox.jsx]: _product decision_

File: app/components/inbox/Notification.jsx [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+		const DATE_FORMAT_STRING = 'YYYY-MM-DD HH:mm';
+		let parsed = moment(datestr, DATE_FORMAT_STRING);
+		const withinOneWeek = moment().add(7, 'days') > parsed;
+		return withinOneWeek ? parsed.calendar() : parsed.format('dddd MMM Do');
+	}
+	conjugateNotification = (notification, event) => {
+		const userLink = (
+			<Link to={`/user/${notification.username}`}>{notification.username}</Link>
+		);
+		switch(notification.type) {
```
Comment (tgoodwin) [inline app/components/inbox/Notification.jsx]: this part... uhhhh lol 
needed a switch-style thing that could return jsx code for links

File: app/components/modals/Modal.jsx [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+	}
+	getModal(modalName) {
+		let modalShown;
+		switch(modalName) {
+			case MODAL_TYPES.POST_OPTIONS_MODAL:
+				modalShown = (<PostOptionsModalContainer />);
+				break;
+			case MODAL_TYPES.AUTHOR_OPTIONS_MODAL:
+				modalShown = (<AuthorOptionsModalContainer />);
+			// do not put a default, it will always get reset to this
```
Comment (tgoodwin) [inline app/components/modals/Modal.jsx]: 🤷‍♀️

File: app/components/post/Post.jsx [hunk 9]
```diff
@@ -4,19 +4,20 @@ import { Link } from 'react-router-dom';
...
+import PostHeader from './PostHeader.jsx';
 import PostStats from './PostStats.jsx';
 import Comment from './Comment.jsx';
 import AddCommentInput from './AddCommentInput.jsx';
 
 class Post extends React.Component {
 	constructor() {
 		super();
-		this.auth = getAuth();
+		this.auth = getAuth() ;
```
Comment (tgoodwin) [inline app/components/post/Post.jsx]: wow

Comment (tgoodwin) [inline app/components/post/Post.jsx]: dangit

File: app/components/post/PostHeader.jsx [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+						{this.props.username}
+					</Link>
+				</div>
+				<div className='modal-btn' onClick={this.props.toggleModal}>
+					&middot;&middot;&middot;
+				</div>
+			</div>
+		);
+	}
+}
```
Comment (tgoodwin) [inline app/components/post/PostHeader.jsx]: I've been using tabs 😬  but should maybe move to spaces codebase-wide? cause this shit looks weird

File: app/components/post/PostStats.jsx [hunk 11]
```diff
@@ -12,14 +12,14 @@ class PostStats extends React.Component {
 	}
 	render() {
 		// grammar lol
-		const likesNoun = this.props.likeCount == 1 ? 'like' : 'likes';
-		const likeBtnText = this.props.isLiked ? 'Liked' : 'Like';
+		const likesNoun = this.props.likeCount == 1 ? 'star' : 'stars';
+		const likeBtnText = this.props.isLiked ? 'starred' : 'star';
```
Comment (tgoodwin) [inline app/components/post/PostStats.jsx]: language is hard

File: app/components/App.jsx (hunk 1 above)
Comment (cameronmcinnes) [reply app/components/App.jsx]: I think `Promise.all` is fine. We could as define a redux action creator that would do both of these and then call that here. But I think we should probably move loading out of local state and into the UI reducer. Then we don't even need to use promises here. And can just call the two functions.

File: app/components/header/Header.jsx (hunk 4 above)
Comment (cameronmcinnes) [reply app/components/header/Header.jsx]: its a feature!

File: app/components/header/Header.jsx [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+	const logInBtn = (<NavLink to='/login'>Log in</NavLink>);
+	const inboxHeader = (
+		<InboxHeader inviteCount={3} />
+	);
+	const defaultHeader = (
+		<div className='nav-bar-container'>
+			<header className='inline-centered'>
+				{ loggedIn ? logOutBtn : logInBtn }
+				<Link to='/' ><h1 className='logo'>HAPPENINGS</h1></Link>
+				{ loggedIn ? <Link to='/inbox' className='notification-count'>{unreadNotificationCount}</Link> : null }
```
Comment (cameronmcinnes) [inline app/components/header/Header.jsx]: you should be able to just do `{ loggedIn && <Link to='/inbox' className='notification-count'>{unreadNotificationCount}</Link> }`

File: app/components/header/Header.jsx (hunk 5 above)
Comment (cameronmcinnes) [reply app/components/header/Header.jsx]: prob should break out the `inline-center` header and `main-nav` nav into their own components

File: app/components/post/PostHeader.jsx (hunk 10 above)
Comment (cameronmcinnes) [reply app/components/post/PostHeader.jsx]: yea its so wide wtf

Comment (cameronmcinnes) [reply app/components/post/PostHeader.jsx]: https://atom.io/packages/tabs-to-spaces
this is for atom but i assume theres something for sublime also ?

File: app/actions/inbox.js (hunk 2 above)
Comment (cameronmcinnes) [reply app/actions/inbox.js]: `[... new Set(notifications.map(notif => notif.postId)]`

File: app/components/inbox/Notification.jsx [hunk 13]
```diff
@@ -0,0 +1,124 @@
+import React from 'react';
+import classNames from 'classnames';
+import moment from 'moment';
+import { Link } from 'react-router-dom';
+import * as NotificationTypes from '../../constants/notificationTypes';
+import { updateAttending } from '../../utils/post';
+import FollowNotification from './FollowNotification';
+
+class Notification extends React.Component {
+	componentWillMount() {
```
Comment (cameronmcinnes) [inline app/components/inbox/Notification.jsx]: think we should use `componenetDidMount`. `componentWillMount` behaves weirdly.

File: app/components/post/Post.jsx [hunk 14]
```diff
@@ -4,19 +4,20 @@ import { Link } from 'react-router-dom';
 
 import { getAuth } from '../../utils/auth';
 import { toArray as _toArray } from '../../utils/index';
-import { fetchComments, registerUserToLike, addComment, subscribeToComments,
-	registerForCommentsCount, updateLike as _updateLike, deletePost as _deletePost,
-	updateAttending, registerUserAttendance, deleteComment as _deleteComment
+import { fetchComments, addComment, subscribeToComments,
```
Comment (cameronmcinnes) [inline app/components/post/Post.jsx]: lets just import this whole file as `utils` or something and then use the functions like `util.deletePost` etc so we don't have alias and keep track

File: app/components/menus/SearchDropDown.jsx [hunk 15]
```diff
@@ -35,9 +35,10 @@ class SearchResultDropdown extends React.Component {
         </li>
       </Link>
     ));
+    console.log(postResults);
 
     const postList = postResults.map((post, idx) => (
-      <Link to={ `/posts/${post.id}` } key={ idx }>
+      <Link to={ `/events/${idx}` } key={ idx }>
```
Comment (cameronmcinnes) [inline app/components/menus/SearchDropDown.jsx]: should be `/post.id`

File: app/reducers/selectors.js [hunk 16]
```diff
@@ -55,8 +55,8 @@ export const selectPostResults = state => (
 
 export const countUnreadNotifications = state => {
   if (state.session.currentUser && !state.ui.loading.notificationsLoading) {
-    var count = Object.keys(state.entities.inbox.notifications).filter(
-      notification => state.entities.inbox.notifications[notification].read == false
+    var count = Object.keys(state.entities.notifications).filter(
```
Comment (cameronmcinnes) [inline app/reducers/selectors.js]: prob should do `const` instead of `var` and stick w es6

File: package.json [hunk 17]
```diff
@@ -35,13 +35,15 @@
...
     "react": "^16.3.0",
+    "react-copy-to-clipboard": "^5.0.1",
     "react-dom": "^16.2.0",
     "react-dropzone": "^4.2.9",
     "react-redux": "^5.0.7",
     "react-router-dom": "^4.2.2",
     "redux": "^4.0.0",
     "redux-logger": "^3.0.6",
     "redux-thunk": "^2.2.0",
+    "underscore": "^1.9.1",
```
Comment (cameronmcinnes) [inline package.json]: we stopped usin this right? if so lets get rid of this shiiiiii

**Pull request merged**